pkg/object/bunny.go

```diff
@@ -0,0 +1,193 @@
+/*
+ * JuiceFS, Copyright 2024 Juicedata, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package object
+
+import (
+	"bytes"
+	"io"
+	"math"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	bunnystorage "github.com/l0wl3vel/bunny-storage-go-sdk"
+)
+
+type bunnyClient struct {
+	DefaultObjectStorage
+	client   *bunnystorage.Client
+	endpoint string
+}
+
+// Description of the object storage.
+func (b *bunnyClient) String() string {
+	return b.endpoint
+}
+
+// Limits of the object storage.
+func (b *bunnyClient) Limits() Limits {
+	return Limits{
+		IsSupportMultipartUpload: false,
+		IsSupportUploadPartCopy:  false,
+	}
+}
+
+// Get the data for the given object specified by key.
+func (b *bunnyClient) Get(key string, off int64, limit int64) (io.ReadCloser, error) {
+	if limit == -1 {
+		limit = math.MaxInt64
+	}
+	body, err := b.client.DownloadPartial(key, off, limit+off-1)
+	if err != nil {
+		return nil, err
+	}
+	return io.NopCloser(bytes.NewReader(body)), nil
+}
+
+// Put data read from a reader to an object specified by key.
+func (b *bunnyClient) Put(key string, in io.Reader) error {
+	content, readErr := io.ReadAll(in)
+	if readErr != nil {
+		return readErr
+	}
+	return b.client.Upload(key, content, true)
+}
+
+// Delete a object.
+// Requires a conditional retry, since deleting a directory or file called foo/bar requires two different calls to the Bunny API, which JuiceFS does not do
+// Deleting a directory requires a trailing slash in the key to delete, which JuiceFS does not add to the path, leading to test case failures.
+// We implement a conditional retry here to try deleting the directory if the delete for a key of the passed name fails.
+// Deleting keys that do not exist are expected to not throw an error
+func (b *bunnyClient) Delete(key string) error {
+	if err := b.client.Delete(key, false); err != nil {
+		if err.Error() == "Not Found" {
```

Comment:
This error message is confusing, we should only retry it when the key is ending with `/`.

Reply:
Or call `b.client.Delete(key, strings.HasSuffix("/"))`

Reply:
This workaround is necessary to pass the tests.  The boolean flag on `b.client.Delete()` only appends a / character to the path, which means that setting it at all should not be necessary.

Tests fail on repeated test runs with the following implementation:

```
func (b *bunnyClient) Delete(key string) error {
	if err := b.client.Delete(key, false); err != nil {
		if err.Error() == "Not Found" {
			return nil
		} else {
			return err
		}
	}
	return nil
}
```

Well, apparently some tests do not consistently pass paths with a trailing /. The first run works fine, but on any subsequent run a list test fails because there are empty directories left over from the first test run.

### First run
```
=== RUN   TestBunny
    object_storage_test.go:131: PUT testEncodeFile failed: Unauthorized
    object_storage_test.go:180: out-of-range get: '', but got 0, error: Requested Range Not Satisfiable
    object_storage_test.go:284: list with delimiter is not supported
    object_storage_test.go:326: list with delimiter is not supported
    object_storage_test.go:527: https://storage.bunnycdn.com/juicefs-testunit-test/ does not support multipart upload: not supported
--- PASS: TestBunny (10.97s)
PASS
```
### Object storage directory after the first run
![image](https://github.com/juicedata/juicefs/assets/7373184/33816658-4460-44c8-9047-1ae8c40da6e9)

### Second Run
```
go1.20.13 test -v -tags gluster -run TestBunny github.com/juicedata/juicefs/pkg/object      
=== RUN   TestBunny
    object_storage_test.go:131: PUT testEncodeFile failed: Unauthorized
    object_storage_test.go:180: out-of-range get: '', but got 0, error: Requested Range Not Satisfiable
    object_storage_test.go:222: First key should be test, but got a/
--- FAIL: TestBunny (0.65s)
FAIL
FAIL    github.com/juicedata/juicefs/pkg/object 1.020s
FAIL
```

This workaround seems to be necessary, since it is not guaranteed that JuiceFS passes in the trailing /.

Reply:
lgtm

Reply:
I retested it after replacing ListAll with List and the tests fail again. With or without my deletion workaround there are objects left in the unit-test directory. Subsequent runs will always fail again. Is it possible that testStorage does not handle the cleanup of the unit-test directory on object stores, that allow prefixes containing only empty directories? I would propose cleaning up the root directory of the storage prefix like this:

```
diff --git a/pkg/object/object_storage_test.go b/pkg/object/object_storage_test.go
index cfab5b97..a806c4ef 100644
--- a/pkg/object/object_storage_test.go
+++ b/pkg/object/object_storage_test.go
@@ -120,6 +120,7 @@ func testStorage(t *testing.T, s ObjectStorage) {
        }
        prefix := "unit-test/"
        s = WithPrefix(s, prefix)
+       defer s.Delete("/")
        defer func() {
                if err := s.Delete("test"); err != nil {
                        t.Fatalf("delete failed: %s", err)
```

Reply:
It seems like this problem only appears on Object Storage providers that do support the concept of an empty directory, which S3 style APIs do not support. We would need to adjust the tests to clean up the unit-test directory properly.